onnxruntime/core/graph/graph.cc

```diff
@@ -2055,6 +2056,38 @@ void Graph::RemoveInitializedTensor(const std::string& tensor_name) {
   }
 }
 
+Status Graph::ReplaceInitializedTensor(const ONNX_NAMESPACE::TensorProto& new_initializer) {
+  const auto& initializer_name = new_initializer.name();
+  const auto name_to_initializer_it = name_to_initial_tensor_.find(initializer_name);
+  ORT_RETURN_IF_NOT(name_to_initializer_it != name_to_initial_tensor_.end(),
+                    "Failed to find existing initializer with name ", initializer_name, ".");
+
+  assert(name_to_initializer_it->second);
+  const auto& old_initializer = *(name_to_initializer_it->second);
```

Comment:
name_to_initial_tensor_ is an internal of this class. instead of scattering asserts around where it's used, validate when an entry is added that the value is not null so that that becomes an invariant.